Move new/edit external attachment page to the GOV.UK Design System

davidtrussler opened the pull request:
The work in this PR moves the add/edit external attachment pages to the design system layout. See the [Trello card](https://trello.com/c/UUh0Bjp9/612-move-new-edit-attachment-page-to-the-govuk-design-system) for this work. 

| Current layout | Design System layout |
|---|---|
| ![Screenshot 2022-08-31 at 08 36 52](https://user-images.githubusercontent.com/6080548/187620557-474af5f6-bcd3-4225-98c5-a3b75b315f1a.png) | ![Screenshot 2022-09-01 at 15 02 40](https://user-images.githubusercontent.com/6080548/187934393-8ad589bf-d9da-46c8-8ce3-e367823d3c37.png) |
|  | ![Screenshot 2022-09-01 at 15 08 37](https://user-images.githubusercontent.com/6080548/187934804-4fb1d005-9eec-466d-a1a5-d6b63dcff50c.png) |

nnagewad commented:
Looks good. Just checking is the bottom margin between inputs 50px?

nnagewad commented:
Also, just double check the spacing for the back button and the sub-title. May be good to reference what @DilwoarH has done on the taxonomy page.

davidtrussler commented:
@nnagewad I haven't set any specific spacing on these elements. I'll cross reference with the work that's already been done on this as you suggest in the first instance.

DilwoarH commented on app/views/admin/attachments/edit.html.erb:
```diff
@@ -1,11 +1,16 @@
-    <h1>Editing <%= attachment.attachable_model_name %> <%= attachment.readable_type %> attachment</h1>
+<div class="govuk-grid-row">
+  <section class="govuk-grid-column-two-thirds">
+    <%= render "govuk_publishing_components/components/back_link", href: attachable_attachments_path(attachable) %>
```
If you add it like this, it will add everything you need

DilwoarH replied:
```ruby
<% content_for :back_link do %>
  <%= render "govuk_publishing_components/components/back_link", {
    href: admin_edition_path(@edition)
  } %>
<% end %>
```

davidtrussler replied:
Nice, have done.

DilwoarH commented on app/views/admin/attachments/edit.html.erb:
```diff
@@ -1,11 +1,16 @@
 <% page_title "Editing #{attachment.attachable_model_name} #{attachment.readable_type} attachment" %>
```
You probably want to follow this pattern:

```
<% content_for :page_title, "Editing #{attachment.attachable_model_name} #{attachment.readable_type} attachment" %>
<% content_for :title, "Edit #{attachment.readable_type} attachment" %>
<% content_for :context, "Attachments for #{attachment.attachable_model_name}" %>
```

davidtrussler replied:
👍

DilwoarH commented on app/views/admin/attachments/edit_legacy.html.erb:
```diff
@@ -0,0 +1,14 @@
+<% page_title "Editing #{attachment.attachable_model_name} #{attachment.readable_type} attachment" %>
```
same as https://github.com/alphagov/whitehall/pull/6763/files#r960413584

davidtrussler replied:
👍

davidtrussler commented:
@nnagewad I've updated the spacing between elements on this to be 50px each time. I also updated how the back button and heading are rendered, which is via a standard method across all these pages. I've updated the screenshots to show this. 

The effect of this is to render the relative sizes of the headings (the main one and the ones for each element) now appear to be the wrong size relative to each other. I initially followed the font sizes in the Figma file but could yo confirm what we would expect these to be? Currently they are:

|Element|font-size|Figma font-size|
|---|---|---|
|"Attachments for publication"|24px|27px|
|"Add external attachment"|36px|48px|
|"Title (required)"|36px|36px|

nnagewad commented:
Hey David, regarding the size of the font title, I was using Extra large font size + caption as a reference.
- Link to [GOV.UK Typography](https://design-system.service.gov.uk/styles/typography/#headings-with-captions)

davidtrussler commented:
This was updated in the [last commit](https://github.com/alphagov/whitehall/pull/6763/commits/97cb5b9dab28682e20cfb3589ebd97cda23b5f50) so that input, radio and select elements persist the form values where applicable if the user submits the form with errors. Some of this looks a bit clunky to me and there’s probably a more elegant way of achieving this. 

I was also looking for a way to replicate the current behaviour of highlighting which inputs are causing the errors but was unable to see how that can be done. 

I am wondering if this port to the design system, which effectively bypasses the form builder currently in use is going to be problematic and where we should look at ways of porting it whilst maintaining the form builder.

DilwoarH commented on app/views/admin/attachments/_form.html.erb:
```diff
@@ -1,12 +1,28 @@
-    <%= form.text_field :title %>
+  <% if form.object.errors.any? %>
+    <%= render "govuk_publishing_components/components/error_summary", {
+      title: "To save the #{form.object.class.name.demodulize.underscore.humanize.downcase} please fix the following issues:",
```
Is this the correct variable to use?

DilwoarH replied:
What does it output?

davidtrussler replied:
Good question, I have no idea where that came from. It outputs the type of attachment (e.g. "external"). I've replaced it with what we've used elsewhere i.e. `#{attachment.readable_type}`

DilwoarH commented on app/views/admin/attachments/_form.html.erb:
```diff
@@ -1,12 +1,28 @@
+      name: "attachment[title]",
+      heading_level: 2,
+      heading_size: "l",
+      value: form.object.title.blank? ? "" : form.object.title
```
I think you can just pass in `form.object.title` and it will handle the empty in the component.

```suggestion
      value: form.object.title
```

davidtrussler replied:
👍

DilwoarH commented on app/views/admin/attachments/_form.html.erb:
```diff
@@ -32,12 +48,28 @@
+        name: "attachment[external_url]",
+        heading_level: 2,
+        heading_size: "l",
+        value: form.object.external_url.blank? ? "" : form.object.external_url
```
```suggestion
        value: form.object.external_url
```

davidtrussler replied:
👍

DilwoarH commented on app/views/admin/attachments/_form.html.erb:
```diff
@@ -32,12 +48,28 @@
+      text: "Next"
+    } %>
+  <% else %>
+    <%= form.save_or_cancel cancel: attachable_attachments_path(attachable) %>
```
We should probably replace this with govuk buttons

davidtrussler replied:
I've just kept the existing buttons here for now as it's out of scope for external attachments. This will be replaced when the other forms of attachments are being worked on.

DilwoarH replied:
I think we should add in the styling here because everything else is using govuk design system except this:

```erb
<div class="govuk-button-group">
  <%= render "govuk_publishing_components/components/button", {
    text: "Save"
  } %>

  <a class="govuk-link" href="<%= attachable_attachments_path(attachable) %>">Cancel</a>
</div>
```

DilwoarH commented on app/views/admin/attachments/_reference_fields.html.erb:
```diff
@@ -4,14 +4,71 @@
+    name: "attachment[isbn]",
+    heading_level: 2,
+    heading_size: "l",
+    value: form.object.isbn.blank? ? "" : form.object.isbn
```
```suggestion
    value: form.object.isbn
```

davidtrussler replied:
👍

DilwoarH commented on app/views/admin/attachments/_reference_fields.html.erb:
```diff
@@ -4,14 +4,71 @@
+    name: "attachment[unique_reference]",
+    heading_level: 2,
+    heading_size: "l",
+    value: form.object.unique_reference.blank? ? "" : form.object.unique_reference
```
```suggestion
    value: form.object.unique_reference
```

davidtrussler replied:
👍

DilwoarH commented on features/step_definitions/attachment_steps.rb:
```diff
@@ -26,7 +26,11 @@
 
 When(/^I add an external attachment with the title "(.*?)" and the URL "(.*?)"$/) do |title, url|
-  create_external_attachment(url, title)
+  create_external_attachment(url, title, nil)
+end
+
+When(/^I add an external attachment with the title "(.*?)" and the URL "(.*?)" with design system permission$/) do |title, url|
+  create_external_attachment(url, title, "design_system")
 end
```
instead of a new step, we can just do this:

```
@user.has_permission? "Preview design system"
```

davidtrussler replied:
Ah right, nice!

DilwoarH commented on features/support/attachment_helper.rb:
```diff
@@ -23,18 +23,24 @@ def upload_new_attachment(file_path, attachment_title)
     fill_in "External url", with: url
-    click_on "Save"
+
+    if permissions == "design_system"
+      click_on "Next"
+    else
+      click_on "Save"
+    end
```
```suggestion
    design_system = @user.has_permission? "Preview design system"
    click_on design_system ? "Next" : "Save"
```

davidtrussler replied:
👍

DilwoarH commented on features/support/attachment_helper.rb:
```diff
@@ -23,18 +23,24 @@ def upload_new_attachment(file_path, attachment_title)
   end
 
-  def create_external_attachment(url, attachment_title)
+  def create_external_attachment(url, attachment_title, permissions)
```
```suggestion
  def create_external_attachment(url, attachment_title)
```

davidtrussler replied:
👍

DilwoarH commented on features/support/attachment_helper.rb:
```diff
@@ -23,18 +23,24 @@ def upload_new_attachment(file_path, attachment_title)
     location = current_url
     click_link "Modify attachments"
-    create_external_attachment("http://www.example.com/example", "Example doc")
+    create_external_attachment("http://www.example.com/example", "Example doc", nil)
```
```suggestion
    create_external_attachment("http://www.example.com/example", "Example doc")
```

DilwoarH commented on features/step_definitions/attachment_steps.rb:
```diff
@@ -26,7 +26,11 @@
+  create_external_attachment(url, title, nil)
+end
+
+When(/^I add an external attachment with the title "(.*?)" and the URL "(.*?)" with design system permission$/) do |title, url|
+  create_external_attachment(url, title, "design_system")
```
```suggestion
```

davidtrussler replied:
👍

DilwoarH commented on features/step_definitions/attachment_steps.rb:
```diff
@@ -26,7 +26,11 @@
 
 When(/^I add an external attachment with the title "(.*?)" and the URL "(.*?)"$/) do |title, url|
-  create_external_attachment(url, title)
+  create_external_attachment(url, title, nil)
```
```suggestion
  create_external_attachment(url, title)
```

DilwoarH commented on features/edition-attachments.feature:
```diff
@@ -23,6 +23,26 @@ Feature: Managing attachments on editions
+    When I start editing the attachments from the publication page
+    And I upload a file attachment with the title "Beard Length Statistics 2014" and the file "dft_statistical_data_set_sample.csv"
+    And I upload an html attachment with the title "Beard Length Graphs 2012" and the body "Example **Govspeak body**"
+    And I add an external attachment with the title "Beard Length Illustrations" and the URL "http://www.beardlengths.gov.uk" with design system permission
```
```suggestion
    And I add an external attachment with the title "Beard Length Illustrations" and the URL "http://www.beardlengths.gov.uk"
```

davidtrussler replied:
👍

davidtrussler commented on app/views/admin/attachments/_form.html.erb:
```diff
@@ -32,12 +48,28 @@
 
-  <%= form.save_or_cancel cancel: attachable_attachments_path(attachable) %>
+  <% if attachment.is_a?(ExternalAttachment) %>
+    <%= render "govuk_publishing_components/components/button", {
+      text: "Next"
+    } %>
```
Yes, I believe that's correct: it's the equivalent of the "save" button that we have currently.

davidgisbey commented:
@davidtrussler  I'll whip up an error component that links to the inputs etc. and push it up on a different branch. If you're happy you can just cherry pick the commit 👍

davidtrussler commented:
> @davidtrussler I'll whip up an error component that links to the inputs etc. and push it up on a different branch. If you're happy you can just cherry pick the commit 👍

@davidgisbey Sounds good to me! 👍

davidgisbey commented on app/views/admin/attachments/_reference_fields_legacy.html.erb:
```diff
@@ -0,0 +1,17 @@
+<% if attachable.respond_to?(:translated_locales) && attachable.translated_locales.many? %>
+  <%= form.label :locale, 'Display language' %>
+  <p class="help-block">This determines the translations of the publication that the attachment will appear in.</p>
+  <%= form.select :locale, options_for_locales(attachable.translated_locales), include_blank: 'All languages' %>
```
This section needs to be updated to use the design systems select component

<img width="703" alt="image" src="https://user-images.githubusercontent.com/42515961/188887139-6ee9fb65-c4d0-40a7-b4ab-be1495e7b2b8.png">

DilwoarH commented on app/views/admin/attachments/_form.html.erb:
```diff
@@ -1,12 +1,28 @@
+  <% if form.object.errors.any? %>
+    <%= render "govuk_publishing_components/components/error_summary", {
+      title: "To save the #{attachment.readable_type} attachment please fix the following issues:",
+      items: form.object.errors.full_messages.map do |msg| {
+          text: msg
+        }
+      end
```
Minor styling changes:

```suggestion
      items: form.object.errors.full_messages.map do |msg|
        {
          text: msg
        }
      end
```

davidtrussler replied:
👍

DilwoarH commented on app/views/admin/attachments/_form.html.erb:
```diff
@@ -32,12 +48,28 @@
+    <%= render "govuk_publishing_components/components/button", {
+      text: "Next"
+    } %>
+  <% else %>
```
Did some new logic get added here? Looking at the legacy file, it seems like the save_or_cancel always shows.

davidtrussler replied:
The idea is to keep `save_or_cancel` when the attachment is anything other than an external attachment, in which case it uses the design system `next` button. Either way it submits the form.

DilwoarH commented on app/views/admin/attachments/_hoc_reference_fields.html.erb:
```diff
@@ -1,24 +1,62 @@
+      {
+        value: "0",
+        text: "Numbered",
+        checked: numbered_hoc_paper,
```
I think you can just put this in:

```suggestion
        checked: form.object.numbered_hoc_paper,
```

davidtrussler replied:
Hmm, I tried some variations on that but it didn't persist the state of the radio when the user submits the form with errors. The problem is that `numbered_hoc_paper` is not a valid thing though `unnumbered_hoc_paper` is. I am aware though that setting the values the way it is done here looks pretty clunky and I'm sure there's a more elegant way to deal with it. I just couldn't see what it would be.

davidtrussler replied:
I have now updated this to something that seems to render as required and is somewhat more elegant.

DilwoarH commented on app/views/admin/attachments/_hoc_reference_fields.html.erb:
```diff
@@ -1,24 +1,62 @@
+      {
+        value: "1",
+        text: "Unnumbered",
+        checked: unnumbered_hoc_paper,
```
```suggestion
        checked: form.object.unnumbered_hoc_paper,
```

DilwoarH commented on app/views/admin/attachments/_reference_fields.html.erb:
```diff
@@ -1,17 +1,92 @@
+<%
+  options = [{text: "All languages", value: ""}]
+  options.concat(attachable.translated_locales.map do |locale|
+      {
+        text: native_language_name_for(locale),
+        value: locale
+      }
+    end)
```
```suggestion
  options.concat(attachable.translated_locales.map do |locale|
    {
      text: native_language_name_for(locale),
      value: locale
    }
  end)
```

davidtrussler replied:
👍

DilwoarH commented on features/support/attachment_helper.rb:
```diff
@@ -27,7 +27,10 @@ def create_external_attachment(url, attachment_title)
     fill_in "External url", with: url
-    click_on "Save"
+
+    design_system = @user.has_permission? "Preview design system"
+    click_on design_system ? "Next" : "Save"
```
I think the convention for test files is as follows:

def create_external_attachment(url, attachment_title)
  design_system = @user.has_permission? "Preview design system"

  click_on "Add new external attachment"
  fill_in "Title", with: attachment_title
  fill_in "External url", with: URL
  click_on design_system ? "Next" : "Save"
end

davidtrussler replied:
OK, makes sense 👍

DilwoarH commented:
Hey @davidtrussler - looks good generally, some minor spacing issues but seems to be okay.

Would be useful to jump on a call to see a demo of this.

davidtrussler commented:
> Hey @davidtrussler - looks good generally, some minor spacing issues but seems to be okay.
> 
> Would be useful to jump on a call to see a demo of this.

Yes, a demo would be good. Still need to properly sort out the checked value for the radios.

DilwoarH commented on app/views/admin/attachments/_hoc_reference_fields.html.erb:
```diff
@@ -1,24 +1,49 @@
+    {
+      text: session,
+      value: session,
+      selected: form.object.parliamentary_session == session ? true : false
```
Minor nit pick

```suggestion
      selected: (form.object.parliamentary_session == session)
```

davidtrussler replied:
Yeah that's better 👍

DilwoarH commented on app/views/admin/attachments/_reference_fields.html.erb:
```diff
@@ -1,17 +1,79 @@
+      {
+        value: "0",
+        text: "Numbered",
+        checked: (true if form.object.unnumbered_command_paper.eql?(false)),
```
```suggestion
        checked: form.object.unnumbered_command_paper.eql?(false),
```

DilwoarH reviewed (changes requested):
Looks great - some minor changes but once these are done, happy to approve

davidtrussler commented on app/views/admin/attachments/_hoc_reference_fields.html.erb:
```diff
@@ -1,24 +1,49 @@
+      {
+        value: "0",
+        text: "Numbered",
+        checked: (true if form.object.unnumbered_hoc_paper.eql?(false)),
```
Oh yeah, of course it does. 👍

DilwoarH commented on app/views/admin/attachments/_form.html.erb:
```diff
@@ -32,12 +49,31 @@
+      text: "Next"
+    } %>
+  <% else %>
+    <%= render "govuk_publishing_components/components/button", {
+      text: "Save"
+    } %>
+    <a class="govuk-link" href="attachable_attachments_path(attachable)">Cancel</a>
```
Missing the wrapper

```suggestion
 <div class="govuk-button-group">
    <%= render "govuk_publishing_components/components/button", {
      text: "Save"
    } %>
    <a class="govuk-link" href="attachable_attachments_path(attachable)">Cancel</a>
 </div>
```

davidtrussler commented:
@DilwoarH The error messages have been added to the form fields as below in the last commit. There is also one further commit to persist the values for the numbers for the papers when the form is submitted with errors. 

![Screenshot 2022-09-14 at 15 32 24](https://user-images.githubusercontent.com/6080548/190186092-516d2ea6-2c57-4331-b95b-2e96a5bded1a.png)

davidgisbey commented on app/controllers/admin/attachments_controller.rb:
```diff
@@ -77,6 +84,10 @@ def attachable_attachments_path(attachable)
 private
 
+  def get_layout
+    preview_design_system_user? ? "design_system" : "admin"
```
I think this needs to be scoped to the specific controller actions you're working on or it'll apply to all the controller actions (index etc.)

```

  def get_layout
    return "admin" unless preview_design_system_user?

    case action_name
    when "edit", "update", "new", "create"
      "design_system"
    else
      "admin"
    end
  end
```

davidtrussler replied:
Ah right, makes sense 👍

davidgisbey commented on app/views/admin/attachments/_form.html.erb:
```diff
@@ -1,12 +1,34 @@
-
-  <%= content_tag :fieldset, class: ("right-to-left" if form.object.rtl_locale?) do %>
-    <%= form.text_field :title %>
+  <% if form.object.errors.any? %>
```
sorry i should've said that i merged an error summary component which renders at the top of the page and links to the inputs! My fault. You can call it like this 

```
<% content_for :error_summary, render('shared/error_summary', object: attachment.becomes(Attachment), class_name: @attachment.class.name.demodulize.underscore.humanize.downcase) %>

```

davidgisbey commented on app/views/admin/attachments/_form.html.erb:
```diff
@@ -1,12 +1,34 @@
+      label: {
+        text: "Title (required)"
+      },
+      name: "attachment[title]",
```
So for example here this would need to be 

```
<%= render "govuk_publishing_components/components/input", {
      label: {
        text: "Title (required)"
      },
      name: "attachment[title]",
      id: "attachment[title],
   heading_level: 2,
      heading_size: "l",
      value: form.object.title,
      error_items: (form.object.errors.as_json(full_messages: true)[:title].map do |error|
        {
          "text": error,
        }
      end if form.object.errors[:title].present?)
    } %>
```

for the errors to link

davidgisbey commented on app/views/admin/attachments/_form.html.erb:
```diff
@@ -1,12 +1,34 @@
+      heading_level: 2,
+      heading_size: "l",
+      value: form.object.title,
+      error_items: (form.object.errors.as_json(full_messages: true)[:title].map do |error|
```
I've added a helper method here for this https://github.com/alphagov/whitehall/blob/ed6c5d2a4a4438387f4d8e4cbfc21ffe937e33cf/app/helpers/errors_helper.rb

You can call it by doing the following

```
errors_for(object.errors, :attribute)
```

So in this case it would be 

```
errors_for(attachment.errors, :title)
```

davidgisbey replied:
Yours does something v similar though 👍

davidtrussler replied:
Updated to use the helper.

davidgisbey commented on app/views/admin/attachments/_form.html.erb:
```diff
@@ -32,12 +54,38 @@
+        heading_level: 2,
+        heading_size: "l",
+        value: form.object.external_url,
+        error_items: (form.object.errors.as_json(full_messages: true)[:external_url].map do |error|
```
same as above regarding helper methods for errors and giving it an id so the error links correctly

davidgisbey commented:
Looks great. I've left a few comments around using helpers and the error summary component. Let me know if any of it doesn't make sense 👍

davidtrussler commented:
> Looks great. I've left a few comments around using helpers and the error summary component. Let me know if any of it doesn't make sense or you want to pair on it 👍

Cheers @davidgisbey - that's all clear. The only instance that's a bit problematic is on the `<select>` element, which has a slightly different form to `<input>` elements. I've added it [here](https://github.com/alphagov/whitehall/pull/6763/files#diff-05c5cb654af952bfb6350ee3376d3356b359a7d394b86c50fef228864c63af6cR52) which renders as below. I'm not sure if the error data needs to be passed differently or if the helper needs a tweak. 

![Screenshot 2022-09-15 at 17 44 02](https://user-images.githubusercontent.com/6080548/190460814-2d234443-0c3c-4e47-b7c0-3fd0af4f3153.png)

davidgisbey commented:
You can use the error_for_input helper method for this (should probs be renamed to errors_for_select as that's the only one that can't take an array of hashes.

So here it would be `errors_for_input(attachment.errors, : parliamentary_session)`

davidgisbey commented on app/controllers/admin/attachments_controller.rb:
```diff
@@ -25,6 +28,10 @@ def create
   end
 
+  def edit
+    render(preview_design_system_user? ? "edit" : "edit_legacy")
```
If the fail is unsuccessful it will rerender the edit view here https://github.com/alphagov/whitehall/blob/87af8214f730585c3273613ee4a31f5ad26a06f6/app/controllers/admin/attachments_controller.rb#L44-L46

We will also need this logic 
```
render(preview_design_system_user? ? "edit" : "edit_legacy")
```

there so the correct page is rerendered

davidgisbey commented on app/controllers/admin/attachments_controller.rb:
```diff
@@ -14,7 +15,9 @@ def order
 
-  def new; end
+  def new
+    render(preview_design_system_user? ? "new" : "new_legacy")
```
If the fail is unsuccessful it will rerender the edit view here https://github.com/alphagov/whitehall/blob/87af8214f730585c3273613ee4a31f5ad26a06f6/app/controllers/admin/attachments_controller.rb#L26-L28

We will also need this logic 
```
 render(preview_design_system_user? ? "new" : "new_legacy")```

there so the correct page is rerendered

DilwoarH reviewed (approved):
Looks good - let's wait till Tuesday before we merge as it's a long weekend and Friday afternoon

davidgisbey reviewed (approved):
lgtm
